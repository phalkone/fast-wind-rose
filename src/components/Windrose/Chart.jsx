/* eslint-disable react/jsx-fragments */
import React, { Fragment } from 'react'
import PropTypes from 'prop-types'

/**
 * The chart that can be divided into a specified number of sectors
 */
function Chart (props) {
  return (
    <Fragment>
      {/* Sector lines */}
      {[...new Array(props.sectorCount).keys()].map(i => (
        <Fragment key={i}>
          <line
            x1={props.center}
            x2={props.center}
            y1={10}
            y2={props.center}
            stroke='grey'
            strokeWidth='1'
            transform={`rotate(${(0.5 + i) * (360 / props.sectorCount)},` +
                       ` ${props.center},${props.center})`}
          />
        </Fragment>
      ))}
      {/* Circles */}
      {[...new Array(4).keys()].map(i => (
        <Fragment key={i}>
          <circle
            cx={props.center}
            cy={props.center}
            r={(4 - i) * ((props.center - 10) / 4)}
            fill='none'
            strokeWidth='1'
            stroke='grey'
            opacity={i > 0 ? 0.5 : 1}
          />
        </Fragment>
      ))}
    </Fragment>
  )
}

Chart.propTypes = {
  /**
   * The number of sectors
   */
  sectorCount: PropTypes.number,
  /**
   * The center of the chart. x and y coordinates are the same.
   */
  center: PropTypes.number
}

export default Chart
