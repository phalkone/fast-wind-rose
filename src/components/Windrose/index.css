button {
  position: absolute;
  top: 10px;
  color: rgb(0, 0, 0);
  background-color: white;
  font-weight: bolder;
  border: 1px solid black;
  text-decoration: none;
  padding: 0 5px;
  transition-duration: 0.4s;
}

button:hover {
  background-color: rgb(203, 203, 203);
}

select {
  position: absolute;
  top: 10px;
  left: 10px;
  padding: 2px;
}

svg, select, button {
  font-family: Roboto, "Open Sans", sans-serif;
}
