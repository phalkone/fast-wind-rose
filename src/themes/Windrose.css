.windrose-button{
  position: absolute;
  top: 10px;
  color: rgb(0, 0, 0);
  background-color: white;
  font-weight: bolder;
  border: 1px solid black;
  text-decoration: none;
  padding: 0 5px;
  transition-duration: 0.4s;
}
 
.windrose-button:hover {
  background-color: rgb(203, 203, 203);
}

.windrose-select {
  position: absolute;
  top: 10px;
  left: 10px;
  padding: 2px;
}

.windrose-container {
  font-family: Roboto, "Open Sans", sans-serif;
}