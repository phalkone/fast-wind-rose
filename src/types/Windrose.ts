export interface IWindrose {
  /**
   * Hide or display legend
   */
  legend: boolean,
  /**
   * Width/height of chart. Will be displayed in specified viewbox.
   */
  size: number,
  /**
   * Default number of sectors. Must be included in sectorArray.
   */
  sectorCount: number,
  /**
   * Fixed interval between data points specified in hours. For example if
   * data points are every 30 min, then value should be 0.5
   */
  interval: number,
  /**
   * Scale of speeds with the linked color. Example as follows:
   *  { 0: 'rgb(60,95,156)', 5: 'rgb(94,131,188)' }
   */
  scale: { [n: number]: string },
  /**
   * Array with directional data for the wind
   */
  dirData: object[],
  /**
  * Array with speed data for the wind
  */
  spdData: object[],
  /**
   * Key for speed data
   */
  spdKey: string,
  /**
   * Key for directional data
   */
  dirKey: string,
  /**
   * Scales the compilation scale against the viewbox. Choose a value lower
   * than 1 to scale down and larger than 1 to scale up.
   */
  enlarge: number,
  /**
   * Array that defines the number of sectors the user can choose from.
   */
  sectorArray: number[],
  /**
   * Common key between direction and speed data
   */
  commonKey: string
}

export interface ILegend {
  /**
   * Size of the chart. Width = Height.
   */
  size: number,
  /**
   * Scale of speeds with the linked color. Example as follows:
   *  { 0: 'rgb(60,95,156)', 5: 'rgb(94,131,188)' }
   */
  scale: { [n: number]: string }
}

export interface IShip {
  /**
   * The center of the ship path. x and y coordinates are assumed to be the same.
   */
  center: number
}

export interface IChart extends IShip {
  /**
   * The number of sectors
   */
  sectorSize: number
}

export interface IIntervalLabel extends IChart {
  /**
   * Defines which sector is being drawn. 0 being the first sector.
   */
  sector: number,
  /**
   * Interval between data points specified in hours
   */
  interval: number,
  /**
   * Array with the speeds that are part of this sector
   */
  speeds: number[]
}

export interface ISector extends IIntervalLabel, ILegend {
  /**
   * Total bar length of this sector. The sector with the largest interval will
   * have a bar length equal to the radius.
   */
  barLength: number,
  /**
   * Defines the bar length of 1 unit of interval.
   */
  unit: number,
}

export interface IToolTip {
  /**
   * x coordinate of top left corner of tooltip
   */
  x: number,
  /**
   * y coordinate of top left corner of tooltip
   */
  y: number,
  /**
   * text of tooltip
   */
  text: [string, number],
  /**
   * Scale factor due to viewBox of SVG vs actual size of SVG.
   */
  factor: number
}
