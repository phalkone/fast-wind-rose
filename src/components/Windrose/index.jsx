import React, { Fragment } from 'react'
import Legend from './Legend'
import Chart from './Chart'
import Ship from './Ship'
import Sector from './Sector'
import IntervalLabel from './IntervalLabel'
import PropTypes from 'prop-types'

export default class Windrose extends React.Component {
  constructor (props) {
    super(props)
    const sectors = [...new Array(360 / this.props.sectorSize).keys()].map((el) => [])
    for (let i = 0; i < this.props.data.length; i++) {
      let dir = this.props.data[i].direction + (this.props.sectorSize / 2)
      if (dir > 360) dir -= 360
      const cat = Math.floor(dir / this.props.sectorSize)
      sectors[cat].push(this.props.data[i].speed)
    }
    this.max = sectors.map((el) => el.length).sort((a, b) => b - a)[0]
    this.sectors = sectors
  }

  renderSectors () {
    return (
      this.sectors.map((speeds, i) => (
        <Fragment key={i}>
          {speeds.length &&
            <IntervalLabel
              sector={i}
              interval={speeds.length}
              sectorSize={this.props.sectorSize}
            />}
          {speeds.length &&
            <Sector
              sector={i}
              speeds={speeds}
              sectorSize={this.props.sectorSize}
              barLength={(speeds.length / this.max) * 120}
              interval={120 / this.max}
            />}
        </Fragment>
      ))
    )
  }

  render () {
    return (
      <svg
        version='1.1'
        viewBox='0 0 325 260'
        width={this.props.width}
        height={this.props.height}
      >
        {this.props.legend && <Legend />}
        <Chart sectorSize={this.props.sectorSize} />
        {this.renderSectors()}
        <Ship />
      </svg>
    )
  }
}

Windrose.propTypes = {
  height: PropTypes.number,
  width: PropTypes.number,
  sectorSize: PropTypes.number,
  legend: PropTypes.bool,
  data: PropTypes.array
}
