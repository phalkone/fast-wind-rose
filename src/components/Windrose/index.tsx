import React, { createContext, Fragment, useState } from 'react'
import { Legend } from './Legend'
import { Chart } from './Chart'
import { Ship } from './Ship'
import { Sector } from './Sector'
import { divideBySector } from '../../utils/Windrose.util'
import type { IWindrose, IWindroseContext } from '../../types/Windrose.types'
import '../../themes/Windrose.scss'

/**
 * Draws a windrose for provided relative wind data
 * @param {IWindrose} props Properties of the Windrose as descibed by IWindrose.
 */
const Windrose = (props: IWindrose) => {
  /* Use state to set the number of sectors and show/hide legend */
  const [sectorCount, setSectorCount] = useState<number>(props.sectorCount)
  const [legend, setLegend] = useState<boolean>(props.legend)

  /* Divide data points per sector */
  const sectors = divideBySector(sectorCount, props.dirData, props.dirKey,
    props.spdData, props.spdKey, props.commonKey)

  /* Set the viewBox size compared to the compilation size */
  const compilationSize = 260 / props.enlarge

  /* Add sectorCount if not part of sectorArray */
  if (!props.sectorArray.includes(sectorCount)) {
    props.sectorArray.push(sectorCount)
  }

  return (
    <div className='windrose-container'>
      {/* Selector for the number of sectors */}
      <select
        title='Number of sectors'
        value={sectorCount}
        className='windrose-select'
        onChange={e => { setSectorCount(Number(e.target.value)) }}
      >
        {props.sectorArray.map(s => (
          <Fragment key={s}>
            <option value={s}>{s}</option>
          </Fragment>
        ))}
      </select>
      <button
        title='Show/hide legend'
        className='windrose-button'
        onClick={e => { setLegend(!legend) }}
        style={{
          left: props.size - 30
        }}
      >{legend ? '<<' : '>>'}
      </button>
      {/* Windrose SVG with optional legend */}
      <svg
        version='1.1'
        viewBox={`0 0 ${compilationSize + (legend ? 65 : 0)} ${compilationSize}`}
        width={props.size + (legend ? 65 * (props.size / compilationSize) : 0)}
        height={props.size}
      >
        <WindroseContext.Provider value={{
          center: compilationSize / 2,
          scale: props.scale,
          sectorSize: 360 / sectorCount,
          interval: props.interval,
          size: props.size,
          max: sectors.map((el) => el.length).sort((a, b) => b - a)[0]
        }}>
          {/* Draw legend only if option is given as prop */}
          {legend && <Legend/>}
          <Chart/>
          {/* Draw each sector with interval label */}
          {sectors.map((speeds, i) => (
            speeds.length && <Fragment key={i}>
              <Sector
                sector={i}
                speeds={speeds}
              />
            </Fragment>
          ))}
          {/* Draw ship outline */}
          <Ship/>
        </WindroseContext.Provider>
        {/* Ensures that tooltips are always on top */}
        <use href='#tooltip' fill='black' />
        <use href='#tooltiptext' />
      </svg>
    </div>
  )
}

Windrose.defaultProps = {
  size: 260,
  spdKey: 'value',
  dirKey: 'value',
  commonKey: 'timestamp',
  sectorCount: 12,
  enlarge: 1,
  interval: 1,
  sectorArray: [4, 8, 12, 16, 24, 32, 36],
  scale: {
    0: 'rgb(60,95,156)',
    5: 'rgb(94,131,188)',
    10: 'rgb(143,180,232)',
    15: 'rgb(174,203,214)',
    20: 'rgb(220,226,220)',
    25: 'rgb(251,233,94)',
    30: 'rgb(252,195,67)',
    35: 'rgb(245,136,42)',
    40: 'rgb(242,103,33)',
    45: 'rgb(243,63,29)',
    50: 'rgb(244,36,27)'
  }
}

const defaultContext : IWindroseContext = {
  center: 130,
  scale: Windrose.defaultProps.scale,
  sectorSize: 30,
  interval: 1,
  size: 260,
  max: 24
}

export const WindroseContext = createContext(defaultContext)

export default Windrose
