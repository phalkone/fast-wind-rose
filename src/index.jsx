import React from 'react'
import { render } from 'react-dom'
import Windrose from './components/Windrose'

const data = [
  { timestamp: '2021-05-30T11:00:00Z', direction: 180.40, speed: 55 },
  { timestamp: '2021-05-30T12:00:00Z', direction: 180.78, speed: 30 },
  { timestamp: '2021-05-30T13:00:00Z', direction: 180.87, speed: 26 },
  { timestamp: '2021-05-30T14:00:00Z', direction: 180.83, speed: 27 },
  { timestamp: '2021-05-30T15:00:00Z', direction: 180.41, speed: 16 },
  { timestamp: '2021-05-30T16:00:00Z', direction: 180.54, speed: 40 },
  { timestamp: '2021-05-30T17:00:00Z', direction: 180.12, speed: 35 },
  { timestamp: '2021-05-30T18:00:00Z', direction: 180.86, speed: 34 },
  { timestamp: '2021-05-30T19:00:00Z', direction: 180.74, speed: 33 },
  { timestamp: '2021-05-30T20:00:00Z', direction: 306.97, speed: 10 },
  { timestamp: '2021-05-30T21:00:00Z', direction: 341.33, speed: 5 },
  { timestamp: '2021-05-31T06:00:00Z', direction: 348.06, speed: 11 },
  { timestamp: '2021-05-31T07:00:00Z', direction: 350.02, speed: 16 },
  { timestamp: '2021-05-31T08:00:00Z', direction: 348.02, speed: 25 },
  { timestamp: '2021-05-31T09:00:00Z', direction: 349.20, speed: 27 },
  { timestamp: '2021-05-31T10:00:00Z', direction: 345.29, speed: 28 }]

render(<Windrose width={650} height={520} sectorSize={30} data={data} legend />, document.getElementById('root'))
