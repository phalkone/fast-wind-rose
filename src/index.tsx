import React from 'react'
import { render } from 'react-dom'
import Windrose from './components/Windrose'
import { mockData } from './mocks'

const element = document.createElement('div')

render(
  <Windrose
    dirData={mockData.metrics[1].values}
    spdData={mockData.metrics[0].values}
    interval={1}
    sectorCount={12}
    size={500}
    legend
    ship
  />, element)

document.body.appendChild(element)
