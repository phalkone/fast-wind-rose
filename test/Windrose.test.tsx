import React from 'react'
import { render, unmountComponentAtNode } from 'react-dom'
import { act } from 'react-dom/test-utils'
import { mockData } from '../src/mocks'
import renderer from 'react-test-renderer'
import Windrose from '../src/components/Windrose'

let container : HTMLElement
beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  unmountComponentAtNode(container)
  container.remove()
})

it('changes shows/hides the legend when clicked', () => {
  act(() => {
    render(<Windrose
      dirData={mockData.metrics[1].values}
      spdData={mockData.metrics[0].values}
      interval={1}
      sectorCount={12}
      size={500}
      legend
      ship />, container)
  })

  const svg = document.getElementsByClassName('windrose')[0]
  const button = document.getElementsByClassName('windrose-button')[0]
  expect(button.innerHTML).toBe('&lt;&lt;')

  expect(Array.from(svg.children).filter((el : Element) => el.tagName === 'rect').length).toEqual(11)

  act(() => {
    button.dispatchEvent(new MouseEvent('click', { bubbles: true }))
  })

  expect(Array.from(svg.children).filter((el : Element) => el.tagName === 'rect').length).toEqual(0)
  expect(button.innerHTML).toBe('&gt;&gt;')

  act(() => {
    for (let i = 0; i < 5; i++) {
      button.dispatchEvent(new MouseEvent('click', { bubbles: true }))
    }
  })

  expect(button.innerHTML).toBe('&lt;&lt;')
})

test('Render a windrose', () => {
  const comp = renderer.create(
    <Windrose
      dirData={mockData.metrics[1].values}
      spdData={mockData.metrics[0].values}
      interval={1}
      sectorCount={12}
      size={500}
      legend
      ship
    />
  )
  const tree = comp.toJSON()
  expect(tree).toMatchSnapshot()
})
