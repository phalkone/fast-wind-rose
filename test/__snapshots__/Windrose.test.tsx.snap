// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a windrose 1`] = `
<div
  className="windrose-container"
>
  <select
    className="windrose-select"
    onChange={[Function]}
    title="Number of sectors"
    value={12}
  >
    <option
      value={4}
    >
      4
    </option>
    <option
      value={8}
    >
      8
    </option>
    <option
      value={12}
    >
      12
    </option>
    <option
      value={16}
    >
      16
    </option>
    <option
      value={24}
    >
      24
    </option>
    <option
      value={32}
    >
      32
    </option>
    <option
      value={36}
    >
      36
    </option>
  </select>
  <button
    className="windrose-button"
    onClick={[Function]}
    style={
      Object {
        "left": 470,
      }
    }
    title="Show/hide legend"
  >
    &lt;&lt;
  </button>
  <svg
    className="windrose"
    height={500}
    version="1.1"
    viewBox="0 0 325 260"
    width={625}
  >
    <rect
      fill="rgb(60,95,156)"
      height={20}
      stroke="grey"
      strokeWidth="1"
      width={20}
      x={270}
      y={220}
    />
    <text
      fill="black"
      fontSize={18}
      x={295}
      y={245}
    >
      0
    </text>
    <rect
      fill="rgb(94,131,188)"
      height={20}
      stroke="grey"
      strokeWidth="1"
      width={20}
      x={270}
      y={200}
    />
    <text
      fill="black"
      fontSize={18}
      x={295}
      y={225}
    >
      5
    </text>
    <rect
      fill="rgb(143,180,232)"
      height={20}
      stroke="grey"
      strokeWidth="1"
      width={20}
      x={270}
      y={180}
    />
    <text
      fill="black"
      fontSize={18}
      x={295}
      y={205}
    >
      10
    </text>
    <rect
      fill="rgb(174,203,214)"
      height={20}
      stroke="grey"
      strokeWidth="1"
      width={20}
      x={270}
      y={160}
    />
    <text
      fill="black"
      fontSize={18}
      x={295}
      y={185}
    >
      15
    </text>
    <rect
      fill="rgb(220,226,220)"
      height={20}
      stroke="grey"
      strokeWidth="1"
      width={20}
      x={270}
      y={140}
    />
    <text
      fill="black"
      fontSize={18}
      x={295}
      y={165}
    >
      20
    </text>
    <rect
      fill="rgb(251,233,94)"
      height={20}
      stroke="grey"
      strokeWidth="1"
      width={20}
      x={270}
      y={120}
    />
    <text
      fill="black"
      fontSize={18}
      x={295}
      y={145}
    >
      25
    </text>
    <rect
      fill="rgb(252,195,67)"
      height={20}
      stroke="grey"
      strokeWidth="1"
      width={20}
      x={270}
      y={100}
    />
    <text
      fill="black"
      fontSize={18}
      x={295}
      y={125}
    >
      30
    </text>
    <rect
      fill="rgb(245,136,42)"
      height={20}
      stroke="grey"
      strokeWidth="1"
      width={20}
      x={270}
      y={80}
    />
    <text
      fill="black"
      fontSize={18}
      x={295}
      y={105}
    >
      35
    </text>
    <rect
      fill="rgb(242,103,33)"
      height={20}
      stroke="grey"
      strokeWidth="1"
      width={20}
      x={270}
      y={60}
    />
    <text
      fill="black"
      fontSize={18}
      x={295}
      y={85}
    >
      40
    </text>
    <rect
      fill="rgb(243,63,29)"
      height={20}
      stroke="grey"
      strokeWidth="1"
      width={20}
      x={270}
      y={40}
    />
    <text
      fill="black"
      fontSize={18}
      x={295}
      y={65}
    >
      45
    </text>
    <rect
      fill="rgb(244,36,27)"
      height={20}
      stroke="grey"
      strokeWidth="1"
      width={20}
      x={270}
      y={20}
    />
    <text
      fill="black"
      fontSize={18}
      x={295}
      y={45}
    >
      50&lt;
    </text>
    <text
      fill="black"
      fontSize={18}
      x={295}
      y={25}
    >
      kts
    </text>
    <line
      stroke="grey"
      strokeWidth="1"
      transform="rotate(15,130,130)"
      x1={130}
      x2={130}
      y1={10}
      y2={130}
    />
    <line
      stroke="grey"
      strokeWidth="1"
      transform="rotate(45,130,130)"
      x1={130}
      x2={130}
      y1={10}
      y2={130}
    />
    <line
      stroke="grey"
      strokeWidth="1"
      transform="rotate(75,130,130)"
      x1={130}
      x2={130}
      y1={10}
      y2={130}
    />
    <line
      stroke="grey"
      strokeWidth="1"
      transform="rotate(105,130,130)"
      x1={130}
      x2={130}
      y1={10}
      y2={130}
    />
    <line
      stroke="grey"
      strokeWidth="1"
      transform="rotate(135,130,130)"
      x1={130}
      x2={130}
      y1={10}
      y2={130}
    />
    <line
      stroke="grey"
      strokeWidth="1"
      transform="rotate(165,130,130)"
      x1={130}
      x2={130}
      y1={10}
      y2={130}
    />
    <line
      stroke="grey"
      strokeWidth="1"
      transform="rotate(195,130,130)"
      x1={130}
      x2={130}
      y1={10}
      y2={130}
    />
    <line
      stroke="grey"
      strokeWidth="1"
      transform="rotate(225,130,130)"
      x1={130}
      x2={130}
      y1={10}
      y2={130}
    />
    <line
      stroke="grey"
      strokeWidth="1"
      transform="rotate(255,130,130)"
      x1={130}
      x2={130}
      y1={10}
      y2={130}
    />
    <line
      stroke="grey"
      strokeWidth="1"
      transform="rotate(285,130,130)"
      x1={130}
      x2={130}
      y1={10}
      y2={130}
    />
    <line
      stroke="grey"
      strokeWidth="1"
      transform="rotate(315,130,130)"
      x1={130}
      x2={130}
      y1={10}
      y2={130}
    />
    <line
      stroke="grey"
      strokeWidth="1"
      transform="rotate(345,130,130)"
      x1={130}
      x2={130}
      y1={10}
      y2={130}
    />
    <circle
      cx={130}
      cy={130}
      fill="none"
      opacity={1}
      r={120}
      stroke="grey"
      strokeWidth="1"
    />
    <circle
      cx={130}
      cy={130}
      fill="none"
      opacity={0.5}
      r={90}
      stroke="grey"
      strokeWidth="1"
    />
    <circle
      cx={130}
      cy={130}
      fill="none"
      opacity={0.5}
      r={60}
      stroke="grey"
      strokeWidth="1"
    />
    <circle
      cx={130}
      cy={130}
      fill="none"
      opacity={0.5}
      r={30}
      stroke="grey"
      strokeWidth="1"
    />
    <path
      d="M 161.06 14.09 A 120 120, 0, 0, 0, 98.94 14.09 L 130 130 Z"
      fill="rgb(243,63,29)"
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      transform="rotate(0, 130, 130)"
    />
    <path
      d="M 157.18 28.58 A 105 105, 0, 0, 0, 102.82 28.58 L 130 130 Z"
      fill="rgb(245,136,42)"
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      transform="rotate(0, 130, 130)"
    />
    <path
      d="M 141.65 86.53 A 45 45, 0, 0, 0, 118.35 86.53 L 130 130 Z"
      fill="rgb(143,180,232)"
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      transform="rotate(0, 130, 130)"
    />
    <path
      d="M 100.49 11.62 A 122 122, 0, 0, 1, 159.51 11.62"
      fill="none"
      id="curved0"
      transform="rotate(0, 130, 130)"
    />
    <text
      fill="black"
      fontSize="11px"
      textAnchor="middle"
    >
      <textPath
        href="#curved0"
        startOffset="50%"
      >
        8.0h
      </textPath>
    </text>
    0
    0
    0
    0
    0
    0
    <path
      d="M 133.88 115.51 A 15 15, 0, 0, 0, 126.12 115.51 L 130 130 Z"
      fill="rgb(174,203,214)"
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      transform="rotate(210, 130, 130)"
    />
    <path
      d="M 161.45 3.86 A 130 130, 0, 0, 0, 98.55 3.86"
      fill="none"
      id="curved7"
      transform="rotate(210, 130, 130)"
    />
    <text
      fill="black"
      fontSize="11px"
      textAnchor="middle"
    >
      <textPath
        href="#curved7"
        startOffset="50%"
      >
        1.0h
      </textPath>
    </text>
    <path
      d="M 133.88 115.51 A 15 15, 0, 0, 0, 126.12 115.51 L 130 130 Z"
      fill="rgb(174,203,214)"
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      transform="rotate(240, 130, 130)"
    />
    <path
      d="M 161.45 3.86 A 130 130, 0, 0, 0, 98.55 3.86"
      fill="none"
      id="curved8"
      transform="rotate(240, 130, 130)"
    />
    <text
      fill="black"
      fontSize="11px"
      textAnchor="middle"
    >
      <textPath
        href="#curved8"
        startOffset="50%"
      >
        1.0h
      </textPath>
    </text>
    <path
      d="M 141.65 86.53 A 45 45, 0, 0, 0, 118.35 86.53 L 130 130 Z"
      fill="rgb(174,203,214)"
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      transform="rotate(270, 130, 130)"
    />
    <path
      d="M 100.49 11.62 A 122 122, 0, 0, 1, 159.51 11.62"
      fill="none"
      id="curved9"
      transform="rotate(270, 130, 130)"
    />
    <text
      fill="black"
      fontSize="11px"
      textAnchor="middle"
    >
      <textPath
        href="#curved9"
        startOffset="50%"
      >
        3.0h
      </textPath>
    </text>
    <path
      d="M 149.41 57.56 A 75 75, 0, 0, 0, 110.59 57.56 L 130 130 Z"
      fill="rgb(244,36,27)"
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      transform="rotate(300, 130, 130)"
    />
    <path
      d="M 145.53 72.04 A 60 60, 0, 0, 0, 114.47 72.04 L 130 130 Z"
      fill="rgb(174,203,214)"
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      transform="rotate(300, 130, 130)"
    />
    <path
      d="M 137.76 101.02 A 30 30, 0, 0, 0, 122.24 101.02 L 130 130 Z"
      fill="rgb(143,180,232)"
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      transform="rotate(300, 130, 130)"
    />
    <path
      d="M 100.49 11.62 A 122 122, 0, 0, 1, 159.51 11.62"
      fill="none"
      id="curved10"
      transform="rotate(300, 130, 130)"
    />
    <text
      fill="black"
      fontSize="11px"
      textAnchor="middle"
    >
      <textPath
        href="#curved10"
        startOffset="50%"
      >
        5.0h
      </textPath>
    </text>
    <path
      d="M 153.29 43.07 A 90 90, 0, 0, 0, 106.71 43.07 L 130 130 Z"
      fill="rgb(174,203,214)"
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      transform="rotate(330, 130, 130)"
    />
    <path
      d="M 141.65 86.53 A 45 45, 0, 0, 0, 118.35 86.53 L 130 130 Z"
      fill="rgb(143,180,232)"
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      transform="rotate(330, 130, 130)"
    />
    <path
      d="M 133.88 115.51 A 15 15, 0, 0, 0, 126.12 115.51 L 130 130 Z"
      fill="rgb(60,95,156)"
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      transform="rotate(330, 130, 130)"
    />
    <path
      d="M 100.49 11.62 A 122 122, 0, 0, 1, 159.51 11.62"
      fill="none"
      id="curved11"
      transform="rotate(330, 130, 130)"
    />
    <text
      fill="black"
      fontSize="11px"
      textAnchor="middle"
    >
      <textPath
        href="#curved11"
        startOffset="50%"
      >
        6.0h
      </textPath>
    </text>
    <path
      d="M 120 145 H 140 C 140 145 140 110 130 110 C 120 110 120 145 120 145 Z"
      fill="none"
      opacity="0.7"
      stroke="grey"
      strokeWidth="3"
    />
    <use
      fill="black"
      href="#tooltip"
    />
    <use
      href="#tooltiptext"
    />
  </svg>
</div>
`;
